docs/Technical-Documentation/tech-memos/parsing-refactor/parsing-refactor.md

```diff
@@ -0,0 +1,331 @@
+
+```python
+"""Base parser logic associated to all parser classes."""
+
+from abc import ABC, abstractmethod
+from tdpservice.parsers.row_schema import SchemaManager
+from tdpservice.parsers.util import SortedRecords
+
+class BaseParser(ABC):
+    """Abstract base class for all parsers."""
+
+    def __init__(self, datafile, dfs, section, program_type):
+        super().__init__()
+        self.datafile = datafile
+        self.dfs = dfs
+        self.section = section
+        self.program_type = program_type
+
+        # Since we know the program type and the section of the file. We can explicitely initialize the schema manager.
+        # This would involve making the schema manager "smarter". That is, we need to give the manager it's section and
+        # program type, then it can instantiate all of the correct row schemas and parse more explicitely.
+        self.schema_manager = SchemaManager(datafile, program_type, section)
+
+        # Initialized in child classes since they know how their file is encoded.
+        self.decoder = None
+
+        self.current_line = ""
+        self.current_line_num = 0
+        self.errors = dict()
+
+        # Specifying unsaved_records here may or may not work for FRA files. If not, we can move it down the
+        # inheritance hierarchy.
+        self.unsaved_records = SortedRecords(section)
+        self.unsaved_parser_errors = dict()
+        self.num_errors = 0
+
+    @abstractmethod
+    def parse_and_validate(self):
+        """To be overriden in child class."""
+        # Should have the same return as parse.py::parse_datafile
+        pass
+
+    @abstractmethod
+    def _init_schema_manager(self):
+        """To be overridden in child class."""
+        pass
+
+    def bulk_create_records(self, header_count, flush=False):
+        """Bulk create unsaved_records."""
+        pass
+
+    def bulk_create_errors(self, batch_size=5000, flush=False):
+        """Bulk create unsaved_parser_errors."""
+        pass
+
+    def rollback_records(self):
+        """Delete created records in the event of a failure."""
+        pass
+
+    def rollback_parser_errors(self):
+        """Delete created errors in the event of a failure."""
+        pass
+
+    def create_no_records_created_pre_check_error(self):
+        """Generate a precheck error if no records were created."""
+        pass
+```
+
+If you look at this class and compare it to `parse.py` you should see a lot of similarity! This class aggregates a lot of the current parser's utility functions while cleaning up their interfaces. Most or all of the input parameters to the functions were easily able to be moved to class level members. Using this template, TANF/SSP/Tribal and FRA based parsers can be created. The following templates provide an suggested initial implementation.
```

Comment:
I love the idea of rolling most of the complexity and functions from `parse.py` into this new class to be called where appropriate as I imagine we'll have less branches resulting from if's and other conditionals. We can just overload these functions for specifics in concrete program/datafile types.